### PR TITLE
Add and implement PlayerRecipeBookClickEvent

### LINKED ISSUE #1783: Event for players clicking on item in recipe book
**Is your feature request related to a problem? Please describe.**
There currently isn't an obvious way to detect nor prevent players clicking on an item in the recipe book, which moves items from the crafting slots to the main inventory in the process.

**Describe the solution you'd like**
A new cancellable event (`RecipeBookClickEvent`?) that is fired as a result of receiving [`PacketPlayInAutoRecipe`](https://wiki.vg/Protocol#Craft_Recipe_Request) or in the `AutoRecipe` classes, that can:
* Prevent items being moved around the inventory as a result of this
* (Possibly) Prevent a [response](https://wiki.vg/Protocol#Craft_Recipe_Response) being sent
* (Possibly) Change the recipe ID sent in the response

**Describe alternatives you've considered**
An alternative is a permission to disable this - Spigot had `minecraft.autocraft` briefly, but this was [removed](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/diff/nms-patches/PlayerConnection.patch?until=2a927e8638f66bb076b68ae8a0d1d0ee9ae72732).

Another alternative would be to fire events when the individual items are moved around the inventory, though I'm not sure if there are any existing events that would be suitable for this?

**Additional context**
The lack of any events for this occurred to me while investigating EssentialsX/Essentials#1289, which is a duplication exploit caused by how EssentialsX's `/recipe` command works. The command places the items of the recipe into a crafting table inventory, and the plugin then blocks any inventory clicks until the player closes the inventory to stop players obtaining the items. However, clicking on a recipe book recipe moves all the items to the inventory without firing any events, allowing the player to keep the items that would normally disappear when closing the `/recipe` inventory.

### PR DESCRIPTION
**Pull request by RealFloorIsJava**
This adds and implements the event discussed in #1783. The event can be cancelled, the requested recipe can be changed, and whether or not the maximum amount of results should be attempted can be also changed.

I've also created a simple example plugin using the event: https://github.com/LordKorea/test-recipeplugin

This is my first PR for Paper, so if you need more info / changes to the patch let me know :)

Closes #1783

### CODE REVIEW

File: Spigot-Server-Patches/0399-Implement-PlayerRecipeBookClickEvent.patch [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ 
+@@ -2435,9 +2437,14 @@ public class PlayerConnection implements PacketListenerPlayIn {
+         PlayerConnectionUtils.ensureMainThread(packetplayinautorecipe, this, this.player.getWorldServer());
+         this.player.resetIdleTimer();
+         if (!this.player.isSpectator() && this.player.activeContainer.windowId == packetplayinautorecipe.b() && this.player.activeContainer.c(this.player) && this.player.activeContainer instanceof ContainerRecipeBook) {
+-            this.minecraftServer.getCraftingManager().a(packetplayinautorecipe.c()).ifPresent((irecipe) -> {
+-                ((ContainerRecipeBook) this.player.activeContainer).a(packetplayinautorecipe.d(), irecipe, this.player);
+-            });
++            // Paper start - fire event for clicking recipes in the recipe book
++            PlayerRecipeBookClickEvent event = new PlayerRecipeBookClickEvent(player.getBukkitEntity(), CraftNamespacedKey.fromMinecraft(packetplayinautorecipe.c()), packetplayinautorecipe.d());
```
Comment (jacobsandersen) [inline Spigot-Server-Patches/0399-Implement-PlayerRecipeBookClickEvent.patch]: Maybe avoid the import by specifying the fully qualified class name for the event & CraftNamespacedKey

Comment (RealFloorIsJava) [reply Spigot-Server-Patches/0399-Implement-PlayerRecipeBookClickEvent.patch]: thanks, did that :)

**Comment by LoneDev6** [comment]
When will this be included in official paper spigot?

**Comment by pop4959** [comment]
This would be extremely useful. :smile:

**Review by jacobsandersen — APPROVED** [review]

**Comment by zachbr** [comment]
// TODO

**Pull request closed without merging**